algorithms.py

```diff
@@ -228,6 +232,90 @@ def add(self, individual):
     def __iter__(self):
         return iter(self._ancestors)
 
+    def find_indv(self, indv):
+        """
+        find the index of an ancestor in population
+        """
+        return self._ancestors.index(indv)
+
+
+class TrajectorySimulator(object):
+    """
+    Class to simulate an allele frequency trajectory on which to condition
+    the coalescent simulation.
+    """
+    def __init__(self, initial_freq, end_freq, alpha):
+        self._allele_freqs = [end_freq]
+        self._times = [0.]
+        self._start_time = 0
+        self._start_size = 1.0
+        self._growth_rate = 0
+        self._initial_freq = initial_freq
+        self._end_freq = end_freq
+        self._alpha = alpha
+        self._initial_dt = 1e-06
+        self._delta_t_mod = 1.0 / 40.0
+
+    def _genic_selection_stochastic_forwards(self, dt, current_freq, alpha):
+        ux = (alpha * current_freq * (1 - current_freq)) / np.tanh(alpha * current_freq)
+        if random.random() < 0.5:
+            current_freq += (ux * dt) + np.sqrt(current_freq * (1.0 - current_freq) * dt)
+        else:
+            current_freq += (ux * dt) - np.sqrt(current_freq * (1.0 - current_freq) * dt)
+        return current_freq
+
+    def _simulate(self):
+        """
+        proposes a sweep trajectory
+        returns the acceptance probability
+        given series of population sizes
+        """
+        ttau = 0
+        x = self._end_freq  # backward time
+        t_inc = self._initial_dt  # will fix to account for changing popnsize
+        current_size = self._start_size
+        N_max = self._start_size
+
+        # NOTE JK: I removed the "events" argument from the Trajectory class here
+        # because it seems like an unnecessary complexity for this initial prototype and
+        # I don't understand what happens in the single population size regime. What's
+        # the simplest possible instance of this trajectory that we can use as input
+        # to the coalescent simulation phase?
+        # I've added the _modifier_events variable here so that the code runs.
+        self._modifier_events = []
+
+        # Go through each event looking for population size changes in popn0
+        # not dealing with sweeps in other popns or growth
+        for event in self._modifier_events:
```

Comment:
Rather than passing in events, it would be cleaner here to pass in the actual sequence of population sizes. That way this class won't need to know anything about the events and it'll be easier to test. Does this make sense?